mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -0,0 +1,217 @@
+    partitioner = part
+  }
+
+  private lazy val dims: (Long, Long) = getDim
+
+  override def numRows(): Long = dims._1
+  override def numCols(): Long = dims._2
+
+  /** Returns the dimensions of the matrix. */
+  def getDim: (Long, Long) = {
+    // picks the sizes of the matrix with the maximum indices
+    def pickSizeByGreaterIndex(
+        example: (Int, Int, Int, Int),
+        base: (Int, Int, Int, Int)): (Int, Int, Int, Int) = {
+      if (example._1 > base._1 && example._2 > base._2) {
+        (example._1, example._2, example._3, example._4)
+      } else if (example._1 > base._1) {
+        (example._1, base._2, example._3, base._4)
+      } else if (example._2 > base._2) {
+        (base._1, example._2, base._3, example._4)
+      } else {
+        (base._1, base._2, base._3, base._4)
+      }
+    }
+
+    val lastRowCol = rdd.treeAggregate((0, 0, 0, 0))(
+      seqOp = (c, v) => (c, v) match { case (base, ((blockXInd, blockYInd), mat)) =>
+        pickSizeByGreaterIndex((blockXInd, blockYInd, mat.numRows, mat.numCols), base)
+      },
+      combOp = (c1, c2) => (c1, c2) match {
+        case (res1, res2) =>
+          pickSizeByGreaterIndex(res1, res2)
+      })
+
+    (lastRowCol._1.toLong * partitioner.rowPerBlock + lastRowCol._3,
+      lastRowCol._2.toLong * partitioner.colPerBlock + lastRowCol._4)
+  }
+
+  /** Returns the Frobenius Norm of the matrix */
+  def normFro(): Double = {
+    math.sqrt(rdd.map { mat => mat._2 match {
+      case sparse: SparseMatrix =>
+        sparse.values.map(x => math.pow(x, 2)).sum
+      case dense: DenseMatrix =>
+        dense.values.map(x => math.pow(x, 2)).sum
+    }
+    }.reduce(_ + _))
+  }
+
+  /** Cache the underlying RDD. */
+  def cache(): DistributedMatrix = {
+    rdd.cache()
+    this
+  }
+
+  /** Set the storage level for the underlying RDD. */
+  def persist(storageLevel: StorageLevel): DistributedMatrix = {
+    rdd.persist(storageLevel)
+    this
+  }
+
+  /** Collect the distributed matrix on the driver as a local matrix. */
+  def toLocalMatrix(): Matrix = {
+    val parts = rdd.collect().sortBy(x => (x._1._2, x._1._1))
+    val nRows = numRows().toInt
+    val nCols = numCols().toInt
+    val values = new Array[Double](nRows * nCols)
+
+    parts.foreach { part =>
+      val rowOffset = part._1._1 * partitioner.rowPerBlock
+      val colOffset = part._1._2 * partitioner.colPerBlock
+      val block = part._2
+      var j = 0
+      while (j < block.numCols) {
+        var i = 0
+        val indStart = (j + colOffset) * nRows + rowOffset
+        val indEnd = block.numRows
+        val matStart = j * block.numRows
+        val mat = block.toArray
+        while (i < indEnd) {
+          values(indStart + i) = mat(matStart + i)
+          i += 1
+        }
+        j += 1
+      }
+    }
+    new DenseMatrix(nRows, nCols, values)
+  }
+
+  /** Collects data and assembles a local dense breeze matrix (for test only). */
+  private[mllib] def toBreeze(): BDM[Double] = {
```

Comment:
Oh I see.  Yeah, I think we should change that later, but later is fine since it's internal.